packages/components/src/components/editor/graph/data/useCustomGraph.ts

```diff
@@ -0,0 +1,154 @@
+import { useCallback, useEffect, useState } from 'react';
+import {
+  addEdge,
+  applyEdgeChanges,
+  applyNodeChanges,
+  MarkerType,
+  useReactFlow,
+  type Edge,
+  type Node,
+  type OnConnect,
+  type OnEdgesChange,
+  type OnNodesChange
+} from '@xyflow/react';
+import type { CustomNodeData, GraphNode, GraphProps } from '@/components/editor/graph/graph';
+import { getLayoutedElements, type Direction } from '@/components/editor/graph/data/getLayoutedElements';
+
+const useCustomGraph = ({ graphNodes, options }: GraphProps) => {
+  const [nodes, setNodes] = useState<Node[]>([]);
+  const [edges, setEdges] = useState<Edge[]>([]);
+  const [selectedNode, setSelectedNode] = useState<string>('all');
+
+  const { getNodes, getEdges, fitView } = useReactFlow<GraphNode>();
+
+  const onNodesChange: OnNodesChange = useCallback(changes => setNodes(nds => applyNodeChanges(changes, nds)), [setNodes]);
+  const onEdgesChange: OnEdgesChange = useCallback(changes => setEdges(eds => applyEdgeChanges(changes, eds)), [setEdges]);
+
+  const onConnect: OnConnect = useCallback(
+    params =>
+      setEdges(eds =>
+        addEdge(
+          {
+            ...params,
+            type: 'floating',
+            markerEnd: { type: MarkerType.Arrow }
+          },
+          eds
+        )
+      ),
+    [setEdges]
+  );
+
+  const onLayout = useCallback(
+    (direction: Direction) => {
+      const { nodes: layoutedNodes, edges: layoutedEdges } = getLayoutedElements({ nodes: getNodes(), edges: getEdges(), direction });
+      setNodes([...layoutedNodes]);
+      setEdges([...layoutedEdges]);
+      window.requestAnimationFrame(() => {
+        fitView();
+      });
+    },
+    [fitView, getEdges, getNodes]
+  );
+
+  const onFilterApply = useCallback(
+    (selectedId?: string) => {
+      setSelectedNode(selectedId ?? 'all');
+      const filteredNodes = options?.filter && selectedId ? filterNodesBySelectedNode(graphNodes, selectedId) : graphNodes;
+      const { nodes: layoutedNodes, edges: layoutedEdges } = getLayoutedElements({
+        ...mapNodesAndEdges(filteredNodes, getNodes(), options, selectedId),
+        direction: 'TB'
+      });
+      setNodes([...layoutedNodes]);
+      setEdges([...layoutedEdges]);
+      setTimeout(() => {
+        window.requestAnimationFrame(() => {
+          fitView();
+        });
+      }, 0);
+    },
+    [fitView, graphNodes, options, getNodes]
+  );
+
+  // Initialize layouted nodes and edges
+  useEffect(() => {
+    setSelectedNode('all');
+    setNodes(getLayoutedElements({ ...mapNodesAndEdges(graphNodes, getNodes(), options), direction: 'TB' }).nodes);
+    setEdges(getLayoutedElements({ ...mapNodesAndEdges(graphNodes, getNodes(), options), direction: 'TB' }).edges);
+  }, [graphNodes, options, getNodes]);
+
+  return { nodes, edges, selectedNode, onNodesChange, onEdgesChange, onConnect, onLayout, onFilterApply };
+};
+
+export default useCustomGraph;
```

Comment:
I think I do not use anywhere default exports in this repo, I would directly export the hook, so it is named.